File: test/frontend/test_results.pl
use saliweb::Test;
use Test::More 'no_plan';
use Test::Exception;
use File::Temp qw(tempdir);

BEGIN {
    use_ok('cryptosite');
}

my $t = new saliweb::Test('cryptosite');

# Check results page

# Check display_ok_job
{
    my $frontend = $t->make_frontend();
    my $job = new saliweb::frontend::CompletedJob($frontend,
                        {name=>'testjob', passwd=>'foo', directory=>'/foo/bar',
                         archive_time=>'2009-01-01 08:45:00'});
    my $ret = $frontend->display_ok_job($frontend->{CGI}, $job);
    like($ret, '/Job.*testjob.*has completed.*chimera\.zip.*' .
               'UCSF Chimera session file/ms', 'display_ok_job');
}

# Check display_failed_job at DONE stage
{
    my $frontend = $t->make_frontend();
    my $job = new saliweb::frontend::CompletedJob($frontend,
                        {name=>'testjob', passwd=>'foo', directory=>'/foo/bar',
                         archive_time=>'2009-01-01 08:45:00'});
    my $ret = $frontend->display_failed_job($frontend->{CGI}, $job, "DONE");
    like($ret, '/Your CryptoSite job.*testjob.*failed to produce any ' .
               'prediction.*' .
               'please see the.*#errors.*help page.*For more information, ' .
               'you can.*framework\.log.*download the CryptoSite file\-check ' .
               'log file.*' .
               'contact us/ms',
         'display_failed_job_done');
}

# Check display_failed_job at setup stage
{
    my $frontend = $t->make_frontend();
    my $job = new saliweb::frontend::CompletedJob($frontend,
                        {name=>'testjob', passwd=>'foo', directory=>'/foo/bar',
                         archive_time=>'2009-01-01 08:45:00'});
    my $ret = $frontend->display_failed_job($frontend->{CGI}, $job,
                                            "pre-AllosMod");
    like($ret, '/Your CryptoSite job.*testjob.*failed to produce any ' .
               'prediction.*' .
               'please see the.*#errors.*help page.*For more information, ' .
               'you can.*setup\.log.*download the CryptoSite setup ' .
               'log file.*' .
               'contact us/ms',
         'display_failed_job_setup');
}

# Check get_results_page
{
    my $frontend = $t->make_frontend();
    my $job = new saliweb::frontend::CompletedJob($frontend,
                        {name=>'testjob', passwd=>'foo', directory=>'/foo/bar',
                         archive_time=>'2009-01-01 08:45:00'});
    my $tmpdir = tempdir(CLEANUP=>1);
    ok(chdir($tmpdir), "chdir into tempdir");

    ok(open(FH, "> stage.out"), "Open stage.out");
    print FH "DONE\n";
    ok(close(FH), "Close stage.out");

    my $ret = $frontend->get_results_page($job);
    like($ret,
         '/Your CryptoSite job.*testjob.*failed to produce any prediction/',
         'get_results_page (failed job)');

    ok(open(FH, "> cryptosite.pol.pred.pdb"), "Open cryptosite.pol.pred.pdb");
    ok(close(FH), "Close cryptosite.pol.pred.pdb");

    $ret = $frontend->get_results_page($job);
    like($ret, '/Job.*testjob.*has completed/',
         '                 (successful job)');

    chdir("/");
}

File: test/frontend/test_results.py
import unittest
import re
import saliweb.test

# Import the cryptosite frontend with mocks
cryptosite = saliweb.test.import_mocked_frontend("cryptosite", __file__,
                                                 '../../frontend')


class Tests(saliweb.test.TestCase):

    def test_results_file(self):
        """Test download of results files"""
        with saliweb.test.make_frontend_job('testjob') as j:
            j.make_file('test.log')
            c = cryptosite.app.test_client()
            rv = c.get('/job/testjob/test.log?passwd=%s' % j.passwd)
            self.assertEqual(rv.status_code, 200)

    def test_ok_job(self):
        """Test display of OK job"""
        with saliweb.test.make_frontend_job('testjob2') as j:
            j.make_file("cryptosite.pol.pred.pdb")
            c = cryptosite.app.test_client()
            for endpoint in ('job', 'results.cgi'):
                rv = c.get('/%s/testjob2?passwd=%s' % (endpoint, j.passwd))
                r = re.compile(rb'Job.*testjob2.*has completed.*chimera\.zip.*'
                               b'UCSF Chimera session file',
                               re.MULTILINE | re.DOTALL)
                self.assertRegexpMatches(rv.data, r)

    def test_failed_job_done(self):
        """Test display of job failed at DONE stage"""
        with saliweb.test.make_frontend_job('testjob3') as j:
            j.make_file("stage.out", "DONE\n\n")
            c = cryptosite.app.test_client()
            rv = c.get('/job/testjob3?passwd=%s' % j.passwd)
            r = re.compile(
                b'Your CryptoSite job.*testjob3.*failed to produce any '
                b'prediction.*'
                b'please see the.*#errors.*help page.*For more information, '
                rb'you can.*framework\.log.*download the CryptoSite '
                rb'file\-check log file.*contact us', re.MULTILINE | re.DOTALL)
            self.assertRegexpMatches(rv.data, r)

    def test_failed_job_setup(self):
        """Test display of job failed at setup stage"""
        with saliweb.test.make_frontend_job('testjob4') as j:
            j.make_file("stage.out", "pre-AllosMod\n")
            c = cryptosite.app.test_client()
            rv = c.get('/job/testjob4?passwd=%s' % j.passwd)
            r = re.compile(
                b'Your CryptoSite job.*testjob.*failed to produce any '
                b'prediction.*'
                b'please see the.*#errors.*help page.*For more information, '
                rb'you can.*setup\.log.*download the CryptoSite setup '
                b'log file.*contact us', re.MULTILINE | re.DOTALL)
            self.assertRegexpMatches(rv.data, r)


if __name__ == '__main__':
    unittest.main()
